document/json/decoder.go

```diff
@@ -0,0 +1,333 @@
+package json
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"reflect"
+
+	"github.com/aws/smithy-go/document"
+	"github.com/aws/smithy-go/document/internal/serde"
+)
+
+type DecoderOptions struct{}
+
+type Decoder struct {
```

Comment:
Why is this needed as opposed to the default json unmarshaler?

Reply:
While document types are JSON-like we want to decouple them from the standard library's expectations and behaviors around the types it supports. (For example cases where Smithy has chosen not to support like Time serialization and blob serialization). By explicitly defining our own Encoder/Decoder interface and struct field tags we will be to evolve the interfaces further when modeled document types are added at a later date. Additionally, this enables us to have tighter control over type usage, such as being able to block usage of generated Smithy types as their document representation could change at a later date when modeled type support is added.